1327 anonymize answer counters

He3lixxx opened the pull request:
Well, if you really want to make sure everything is correct in the end, this takes a bit more than what we initially required thought.

I'd understand if we said that this is too much code so we do what @janno42 wanted to do initally (set to random integer between 0 and `answers_left`, then decrease `answers_left`).

We could remove the single result check. I left it in for now because I ran in exactly that problem (giving answers to NO_ANSWER which then is a invalid database state. Our asserts really saved lifes here :D)

fixes #1327

Edit: We should also note that this generates very small variance, almost every evaluation will end up between 2.5 and 3.5

janno42 commented on evap/evaluation/management/commands/anonymize.py:
```diff
@@ -165,16 +165,15 @@ def anonymize_courses(self):
             random.shuffle(evaluations)
-            self.stdout.write("Anonymizing " + str(len(evaluations)) + " evaluations of semester " + str(semester) +
-                "...")
+            self.stdout.write(f"Anonymizing {len(evaluations)} evaluations of semester {semester}....")
```
three dots would be sufficient :)

He3lixxx commented on evap/evaluation/management/commands/anonymize.py:
```diff
@@ -197,23 +201,75 @@ def anonymize_evaluations(self):
+
+                    generated_counts = [random.random() for c in counters]
+                    generated_sum = sum(generated_counts)
+                    generated_counts = [round(count / generated_sum * original_sum) for count in generated_counts]
```
I think floow + adding up is less code and easier to understand, so will probably change that.

He3lixxx commented on evap/evaluation/management/commands/anonymize.py:
```diff
@@ -197,23 +201,75 @@ def anonymize_evaluations(self):
+                        to_add = original_sum - generated_sum
+                        index = random.randint(0, len(generated_counts) - 1)
+
+                        if to_add < 0:  # don't go below zero
```
This if is unneeded, but will get removed anyway

He3lixxx commented:
@karyon @janno42 @felixrindt
- Changed the `round` part to a `floor` as distribution the remaining answers is then easier
- Added tests

I'd appreciate a review (or a short info if you don't want to review)

janno42 commented on evap/evaluation/management/commands/anonymize.py:
```diff
@@ -197,23 +202,68 @@ def anonymize_evaluations(self):
+
+                    assert original_sum == sum(counter.count for counter in counters)
+
+                if contribution.evaluation.is_single_result:
```
do we need that?

janno42 replied:
just asking because it adds database queries :)

He3lixxx replied:
Ah fair enough. Well, I guess if the test covers it, we can remove it?

karyon commented on evap/evaluation/tests/test_commands.py:
```diff
@@ -1,21 +1,30 @@
-    def test_anonymize_does_not_crash(self, mock_input):
+    @staticmethod
+    def get_random_rating_question_type():
+        return random.choice(list(CHOICES.keys()))
```
it seems uncritical here but generally i would avoid randomness to avoid random failures. not sure what mommy.make allows, but perhaps you can use https://docs.python.org/3/library/itertools.html#itertools.cycle somehow? this would also make sure we're using all the types.

edit: nevermind. the code does not have any conditions based on the question type, it really doesn't matter. same for the other randomness in the other tests.

He3lixxx replied:
I think it's a fair point. It's probably way easier to debug if every execution runs down the same path. I made sure that we always do a `random.seed` before using random so every test run should not always be the same. I also exchanged the question type generation with that `cycle` method, seems like a much better idea there.
